### PR TITLE
feat(flow): use DataFusion's optimizer

### PR DESCRIPTION
**Pull request by discord9**
I hereby agree to the terms of the [GreptimeDB CLA](https://github.com/GreptimeTeam/.github/blob/main/CLA.md).

## What's changed and what's your intention?

use and add a few Datafusion `Analyzer` and `Optimizer` to improve writing sql query in flow:
add three new analyzer in `df_optimizer.rs` file, which is also the major changes happen, other edits are either removal of obsolete functions or adding test for it
- `AvgExpandRule`: rewrite `avg` to `sum/count`
- `TumbleExpandRule`: rewrite `tumble` to `tumble_start` and `tumble_end`
- `CheckGroupByRule`: check if all exprs in group by is also used in select list

also apply `AnalyzerRule::TypeCoercion` to cast type properly
optimizer from Datafusion is also appiled before encode to substrait, include:

- `OptimizeProjections`: identifies and eliminates unused columns
- `CommonSubexprEliminate`: avoid redundant computation of common sub-expressions
- `SimplifyExpressions`: simplifies expressions in the logical plan
- `UnwrapCastInComparison`: rewrites CAST(col) = lit to col = CAST(lit)

## Checklist

- [ ] I have written the necessary rustdoc comments.
- [ ] I have added the necessary unit tests and integration tests.
- [ ] This PR requires documentation updates.

### CODE REVIEW

File: src/flow/src/plan.rs [hunk 1]
```diff
@@ -121,6 +121,8 @@ impl TypedPlan {
 
 /// TODO(discord9): support `TableFunc`（by define FlatMap that map 1 to n)
 /// Plan describe how to transform data in dataflow
+///
+/// This can be considered as a physical plan in dataflow, which describe how to transform data in
```
Comment (waynexia) [inline src/flow/src/plan.rs:125]: Is this sentence finished?

File: src/flow/src/expr/func.rs [hunk 2]
```diff
@@ -571,8 +622,8 @@ impl BinaryFunc {
                     t1 == t2,
                     InvalidQuerySnafu {
                         reason: format!(
-                            "Binary function {:?} requires both arguments to have the same type",
-                            generic
+                            "Binary function {:?} requires both arguments to have the same type, left={:?}, right={:?}",
+                            generic,t1,t2
```
Comment (waynexia) [inline src/flow/src/expr/func.rs:626]: nit:

```suggestion
                            generic, t1, t2
```

File: src/flow/src/expr/func.rs [hunk 3]
```diff
@@ -221,6 +221,57 @@ impl UnaryFunc {
...
+            "tumble_start" | "tumble_end" => {
+                let ts = args.first().context(InvalidQuerySnafu {
+                    reason: "Tumble window function requires a timestamp argument",
+                })?;
+                let window_size = args
+                .get(1)
+                .and_then(|expr| expr.expr.as_literal())
+                .context(InvalidQuerySnafu {
+                    reason: "Tumble window function requires a window size argument"
+                })?.as_string() // TODO(discord9): since df to substrait convertor does not support interval type yet, we need to take a string and cast it to interval instead
```
Comment (waynexia) [inline src/flow/src/expr/func.rs:235]: interval in substrait is supported at present

File: src/flow/src/expr/func.rs [hunk 4]
```diff
@@ -221,6 +221,57 @@ impl UnaryFunc {
...
+                .map(|s|cast(Value::from(s), &ConcreteDataType::interval_month_day_nano_datatype())).transpose().map_err(BoxedError::new).context(
+                    ExternalSnafu
+                )?.and_then(|v|v.as_interval())
+                .with_context(||InvalidQuerySnafu {
+                    reason: format!("Tumble window function requires window size argument to be a string describe a interval, found {:?}", args.get(1))
+                })?;
```
Comment (waynexia) [inline src/flow/src/expr/func.rs:241]: Don't know why rustfmt won't handle this paragraph...

File: src/flow/src/expr/func.rs [hunk 5]
```diff
@@ -221,6 +221,57 @@ impl UnaryFunc {
...
+                            window_size,
+                            start_time,
+                        },
+                        ts.clone(),
+                    ))
+                } else {
+                    unreachable!()
+                }
+            }
+            _ => todo!(),
```
Comment (waynexia) [inline src/flow/src/expr/func.rs:271]: is this reachable?

File: src/flow/src/expr/func.rs (hunk 4 above)
Comment (discord9) [reply src/flow/src/expr/func.rs:241]: > Don't know why rustfmt won't handle this paragraph...

This: https://github.com/rust-lang/rustfmt/issues/3863#issuecomment-700441739
found many other fmt errors too.... `cargo fmt -- --config error_on_line_overflow=true;`

File: src/flow/src/expr/func.rs [hunk 6]
```diff
@@ -347,6 +315,91 @@ impl UnaryFunc {
...
+                                .context(ExternalSnafu)?
+                                .as_datetime()
+                                .context(UnexpectedSnafu {
+                                    reason:
+                                        "Expect start time arg to be datetime after successful cast"
+                                            .to_string(),
+                                })?;
+                            Some(ret)
+                        } else {
+                            None
```
Comment (zhongzc) [inline src/flow/src/expr/func.rs:370]: Is it more appropriate to report an error?

File: src/flow/src/df_optimizer.rs [hunk 7]
```diff
@@ -0,0 +1,599 @@
...
+/// i.e.
+/// ```ignore
+/// proj: avg(x)
+/// -- aggr: [sum(x)/count(x) as avg(x)]
+/// ```
+/// becomes:
+/// ```ignore
+/// proj: sum(x)/count(x) as avg(x)
+/// -- aggr: [sum(x), count(x)]
+/// ```
```
Comment (zhongzc) [inline src/flow/src/df_optimizer.rs:165]: You can do some unitests to see if the results are as expected

File: src/flow/src/df_optimizer.rs [hunk 8]
```diff
@@ -0,0 +1,599 @@
...
+                    Arc::new(aggr_plan),
+                )?);
+                return Ok(Transformed::yes(proj));
+            }
+        }
+    }
+    Ok(Transformed::no(plan))
+}
+
+/// expand `avg(<expr>)` function into `cast(sum((<expr>) AS f64)/count((<expr>)`
```
Comment (zhongzc) [inline src/flow/src/df_optimizer.rs:236]: ditto

File: src/flow/src/df_optimizer.rs [hunk 9]
```diff
@@ -0,0 +1,599 @@
...
+    }
+
+    fn name(&self) -> &str {
+        "tumble_expand"
+    }
+}
+
+/// expand `tumble` in aggr expr to `tumble_start` and `tumble_end`, also expand related alias and column ref
+///
+/// will add `tumble_start` and `tumble_end` to outer projection if not exist before
```
Comment (zhongzc) [inline src/flow/src/df_optimizer.rs:358]: ditto

File: src/flow/src/df_optimizer.rs (hunk 7 above)
Comment (discord9) [reply src/flow/src/df_optimizer.rs:165]: > You can do some unitests to see if the results are as expected

unit test for this is `test_tumble_composite` in `transform/aggr.rs`

**Review by zhongzc — APPROVED** [review]

**Comment by fengjiachun** [comment]
@waynexia PTAL

File: tests/cases/standalone/common/flow/flow_call_df_func.result [hunk 10]
```diff
@@ -224,13 +226,13 @@ admin flush_flow('test_numbers_df_func');
...
-SELECT col_0, col_1 FROM out_num_cnt_df_func;
+SELECT maxmin, time_window FROM out_num_cnt_df_func;
 
-+-------+---------------------+
-| col_0 | col_1               |
-+-------+---------------------+
-| 2     | 2021-07-01T00:00:00 |
-+-------+---------------------+
++--------+---------------------+
+| maxmin | time_window         |
```
Comment (waynexia) [inline tests/cases/standalone/common/flow/flow_call_df_func.result:232]: Is this a breaking change or it's compatible with existing tables

File: src/flow/src/expr/func.rs [hunk 11]
```diff
@@ -347,6 +315,96 @@ impl UnaryFunc {
         }
     }
 
+    pub fn from_tumble_func(name: &str, args: &[TypedExpr]) -> Result<(Self, TypedExpr), Error> {
+        match name {
```
Comment (waynexia) [inline src/flow/src/expr/func.rs:319]: prefer matching `name.to_lowercase()` for wider cases

File: src/flow/src/transform/expr.rs [hunk 12]
```diff
@@ -351,7 +361,13 @@ impl TypedExpr {
                 Ok(TypedExpr::new(ret_expr, ret_type))
             }
             _var => {
-                if VariadicFunc::is_valid_func_name(fn_name) {
+                if fn_name == "tumble_start" || fn_name == "tumble_end" {
```
Comment (waynexia) [inline src/flow/src/transform/expr.rs:364]: I see these literals occur many times. Maybe defining them as constants would be better?

File: tests/cases/standalone/common/flow/flow_call_df_func.result (hunk 10 above)
Comment (discord9) [reply tests/cases/standalone/common/flow/flow_call_df_func.result:232]: > Is this a breaking change or it's compatible with existing tables

compatible, flow will always use existing table info if exist, and only fallback to auto create table when table name and id can't be acquire from `TableSource` manager

**Pull request merged**